<!DOCTYPE html>
<html lang="en">
  <head>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-115753061-1"></script>
    <script>
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'UA-115753061-1');
    </script>


    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 

  </head>

  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li><a href="index.html">Home</a></li>
            <li class="active"><a href="getting_started.html">Getting Started</a></li>
            <li><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>

    </nav>

    <!-- CONTENT -->


    <div class="wrapper">

      <!-- Sidebar Holder -->
      <nav id="sidebar">

        <div class="sidebar-header">
          <h3>Overview</h3>
        </div>

        <ul class="list-unstyled components">
          <li>
            <a href="#Requirements">Requirements</a>
          </li>
          <li>
            <a href="#Installation">Installation</a>
          </li>
          <li>
            <a href="#PreparingData">Preparing the data</a>
          </li>
          <li>
            <a href="#GettingStarted">Getting started</a>
          </li>
          <li>
            <a href="#Output">Output and analysis</a>
          </li>

        </ul>

      </nav>

      <!-- Page Content Holder -->
      <div id="content">



        <div class="container-fluid">
          <div class="row">
            <div class="col-md-10 col-md-offset-1">

              <a class="anchor" name="Requirements"></a>
              <h2> Requirements </h2>
              <b>Relate</b> requires C++11 and cmake version 3.1.<br/>

              If you do not have cmake, you can download pre-compiled binaries and source files of cmake <a href="https://cmake.org/download/">here</a>.

              <a class="anchor" name="Installation"></a>
              <h2> Installation </h2>

              <p> Download <b>Relate</b> from <a href="https://github.com/leospeidel/relate">this git repository</a> to a local directory.<br/>
              Then compile the code: </p>
              <pre><code>cd PATH_TO_RELATE/</b> 
mkdir build</b>
cd build
cmake ..
make</code></pre>
              <p> After compiling the source code, you should find the binaries in <b>bin/</b>.<br/> 

              <a class="anchor" name="PreparingData"></a>
              <h2>Preparing the data</h2>

              <p><b>Relate</b> uses the haps/sample file format (output file format of <a href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html">SHAPEIT2</a>) as input.  </p>
              <ul>
                <li>Please separate your data by chromosome.</li>
                <li>You can <b>convert</b> from <b>vcf</b> to <b>haps/sample</b> using <a href="input_data.html#ConvertFromVcf">this function</a> and from <b>hap/legend/sample</b> to <b>haps/sample</b> using <a href="input_data.html#ConvertFromHapLegendSample">this function</a>.</li>
                <li>We provide a script to prepare your data under
                  <pre><code> PATH_TO_RELATE/scripts/PrepareInputFiles/PrepareInputFiles.sh </code></pre>
                  (see <a href="input_data.html#Prepare">here</a> for how to use the script)<br>
                  This script will
                  <ol>
                    <li>make sure that, for every SNP, the <b>ancestral allele is denoted by 0</b> and the alternative allele is denoted by 1.</li> 
                    <li>make sure that every SNP is <b>biallelic</b>.</li>
                    <li>Optional: filter SNPs (and adjust distances between SNPs) using a genomic mask.</li>
                    <li>Optional: subset the samples in the data set.</li>
                    <li>Optional: generate additional annotation of SNPs that can be appended to the output of <b>Relate</b>.</li>
                  </ol>
                </li>
              </ul>
              <p>You can find a toy data set in the subdirectory <a href="https://github.com/leospeidel/relate"><b>example/</b></a>. </p>


              <a class="anchor" name="GettingStarted"></a>
              <h2>Getting Started</h2>

              <div class="alert alert-warning">
                <strong>Warning!</strong><br/>
                <b>Relate</b> creates temporary files and directories.<br/>
                DO NOT RUN MORE THAN ONE INSTANCE OF <b>RELATE</b> IN THE SAME DIRECTORY.<br/>
                (You can run <b>Relate</b> in different subdirectories.)
              </div>

              <p>If you are not using the example data set, create a new working directory. Then, create a subdirectory called <b>data/</b>.</p>
              <div class="col-xs-offset-0.5 col1">
                <pre><code>mkdir example
cd example
mkdir data</code></pre>
              </div>

              <p>In the subdirectory <b>data/</b>, place the following three files
              <div class="table-responsive">
                <table class="table table-hover">
                  <tbody>
                    <tr>
                      <td>haps</td>
                      <td>A file containing haplotype information (<a href="input_data.html#haps">more info</a>)</td>
                    </tr>
                    <tr>
                      <td>sample</td>
                      <td>A file containing information about individuals in the data set (<a href="input_data.html#sample">more info</a>)</td>
                    </tr>
                    <tr>
                      <td>map</td>
                      <td>A file containing recombination rates and distances (<a href="input_data.html#map">more info</a>)</td>
                    </tr>
                  </tbody>
                </table>
              </div>
              </p>

              <p>
              Now change back to the parent directory <b>example/</b> and execute
              <pre><code> PATH_TO_RELATE/bin/Relate \
      --mode All \
      -m 1.25e-8 \
      -N 30000 \
      --haps data/example.haps \
      --sample data/example.sample \
      --map data/genetic_map.txt \
      -o example</code></pre>
              Output files are <b>example.anc</b> and <b>example.mut</b>. </p>

              <p>
              We are specifying the following arguments:


              <dl class="dl-horizontal">
                <dt>--mode</dt>
                <dd>Mode in which to run <b>Relate</b>. Mode "All" will execute all stages of the algorithm and other modes can be used to execute individual stages (see <a href="documentation.html">here</a>). This is useful, for instance, when parallelizing <b>Relate</b>.</dd>
                <dt>-m,--mutation_rate</dt>
                <dd>Mutation rate per base per generation.</dd>
                <dt>-N,--effective_size</dt>
                <dd>Effective population size of haplotypes. (NOT of individuals! To get the population size of haplotypes, multiply the effective population size of individuals by 2)</dd>
                <dt>--haps</dt>
                <dd>Filename of haps file.</dd>
                <dt>--sample</dt>
                <dd>Filename of sample file.</dd>
                <dt>--map</dt>
                <dd>Filename of genetic map.</dd>
                <dt>-o,--output</dt>
                <dd>Filename of output files without file extension.</dd>
              </dl>


              <b>Relate</b> has a few more optional arguments.

              <dl class="dl-horizontal">
                <dt>--dist</dt>
                <dd>Specifies distances (in BP) between SNPs. This is to correct for regions with low mappability in the genomic mask (A tool that takes a genomic mask and generates a .dist file can be found <a href="input_data.html#FilterHapsUsingMask">here</a>).<</dd>
                <dt>--annot</dt>
                <dd>Add columns to the .mut file. This can also be done after running <b>Relate</b> (a tool for generating a .annot file and/or appends columns to the .mut file can be found <a href="input_data.html#GenerateSNPAnnotations">here</a>). The additional columns are needed, for instance, when estimating population size.</dd>
                <dt>--coal</dt>
                <dd>Specifies variable coalescence rates through time in form of a .coal file. See <a href="modules.html#.coal"> here </a> for file format. Specifying this option will overwrite the --effectiveN option.</dd>
                <dt>--memory</dt>
                <dd>Specifies the approximate memory allowance (in GB) used for storing the distance matrices in memory. <b>Relate</b> will exceed this amount, as it also stores trees and other information in memory; however the option allows to approximately control memory usage. Default is 5GB. <b>Relate</b> can become more efficient (in runtime and hard disk usage) with more memory, particularly for large sample sizes.</dd>
                <dt>--seed</dt>
                <dd>Optional: Seed for random number generator used for branch lengths estimation.</dd>
                <dt>other arguments</dt>
                <dd>Arguments --chunk_index, --first_section, --last_section are only relevant when executing <b>Relate</b> in a mode other that <b>All</b>. This is needed for parallelising <b>Relate</b> (<a href="parallelise.html">more info</a>).</dd>
              </dl>


              <ul>
                <li>Programme options can also be viewed by executing
                  <pre><code> PATH_TO_RELATE/bin/Relate</code></pre></li>
                  <li>Required arguments may differ for each mode. A list of required and optional arguments for mode All can be viewed by executing
                    <pre><code> PATH_TO_RELATE/bin/Relate --mode All</code></pre>
                  </li>
                  <li>All input files can be gzipped.</li>
                  <li>If something goes wrong and <b>Relate</b> terminates with an error, it may leaves behind some temporary files and directories. The directory can be clean-up using
                    <pre><code> PATH_TO_RELATE/bin/Relate \
        --mode Clean \
        -o example</code></pre>
                  </li>
              </ul>


              <a class="anchor" name="Output"></a>
              <h2>Output and analysis</h2>

              <p><b>Relate</b> outputs two files.</p>
              <div class="table-responsive">
                <table class="table table-hover">
                  <tbody>
                    <tr>
                      <td>.anc </td>
                      <td>This file contains the trees. The first line shows the number of haplotypes. The second line shows the number of trees. <br/>
                        Then, each line represents one tree. 
                        <ul>

                          <li> Every line starts with an integer specifying at which SNP the tree begins. </li>
                          <li> Then, the i'th space separated entry shows the parent of the i'th internal node. In brackets we specify the branch length, the number of mutations on this branch, the SNP at which the branch appears, and the SNP at which the branch disappears.</li>
                        </ul></td>
                    </tr>
                    <tr>
                      <td>.mut </td> 
                      <td>This file contains detailed information about each SNP. The header describes each column.<br>
                        Without .annot file, the columns are<br>
                        snp;pos_of_snp;dist;rs-id;tree_index;branch_indices;is_not_mapping;is_flipped;age_begin;age_end;ancestral_allele/alternative_allele;<br>
                        <ol>
                          <li>Number of SNP in .haps file (int)</li>
                          <li>Position of SNP in bp (int)</li>
                          <li>Distance to next SNP in bp (int). If --dist option was specified, this may be different to the actual distance to the next SNP.</li>
                          <li>rd-id of SNP (string)</li>
                          <li>Number of tree, starting from 0, in the .anc file (int)</li>
                          <li>Number of branch on tree. If a mutation cannot be mapped to a unique branch, more than one branch indices are given (space separated ints)</li>
                          <li>0 means SNP could be mapped to a unique branch, 1 means SNP could not be mapped to a unique branch(bool)</li>
                          <li>0 means SNP is not flipped, 1 means SNP is flipped (bool)</li>
                          <li>Age of lower end of branch in generations (dbl)</li>
                          <li>Age of upper end of branch in generations (dbl)</li>
                          <li>Ancestral and alternative alleles at this locus (string)</li>
                        </ol>
                        With .annot fil, the following columns are appended<br>
                        upstream_allele;downstream_allele;pop1;pop2;...
                        <ol start = "12">
                          <li>Allele adjacent to the SNP in the 5' direction</li>
                          <li>Allele adjacent to the SNP in the 3' direction</li>
                          <li>Number of carriers of the derived allele in pop1</li>
                          <li>Number of carriers of the derived allele in pop2</li>
                        </ol>

                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>

          <h4> Using the output </h4>

          <p>We provide a number of add-on modules to analyse the genealogy (look <a href="modules.html">here</a> for details).</p>

          <div class="row">
            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#TreeView">Plot marginal trees</a></h4></div>
                <div class="panel-body">  
                  <p>You can find a bash script under 
                  <pre><code>PATH_TO_RELATE/scripts/TreeView/TreeView.sh</code></pre> 
                  The script will produce a plot of the marginal tree at a bp position of interest with tips labelled by their assigned population.
                  Details about how to run the script can be found <a href="modules.html#TreeView">here</a>)</p>
                </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#CoalescenceRate">Estimate population size</a></h4></div>
                <div class="panel-body">  
                  <p>You can find a bash script under 
                  <pre><code>PATH_TO_RELATE/scripts/EstimatePopulationSize/EstimatePopulationSize.sh</code></pre> 
                  To run the script, all you need are the inferred .anc/.mut files and a .poplabels file (see <a href="input_data.html#poplabels">here</a> for file format). Details about how to run the script can be found <a href="modules.html#CoalescenceRate">here</a>)</p>
                </div>
              </div>
            </div>
          </div>

          <div class="row">
            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#MutationRate">Estimate mutation rates</a></h4></div>
                <div class="panel-body">  
                  <p>You can estimate mutation rates through time using <a href="modules.html#MutationRate">this module</a>. This can be used to calculate average mutation rates, or mutation rates of triplets.</p>
                </div>
              </div>
            </div>
            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Selection">Calculate p-values for selection evidence</a></h4></div>
                <div class="panel-body">  
                  <p>You can calculate a pvalue for selection evidence for every SNP using <a href="modules.html#Selection">this module</a>. The pvalue quantifies whether derived allele has spread surprisingly quickly under the standard coalescent model.</p>
                </div>
              </div>
            </div>
          </div>

          <div class="row">          
            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Extract">Extract trees of interest</a></h4></div>
                <div class="panel-body">  
                  <p>You can extract trees in .newick format, or get genealogies for subpopulations using <a href="modules.html#Extract">this module</a>.</p>
                </div>
              </div>
            </div>
          </div>

        </div>


      </div>
    </div>

    </div>

  </body>
</html>                            
