<!DOCTYPE html>
<html lang="en">

  <head> 

    <script>
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-115753061-1', 'auto');

    // Replace the following lines with the plugins you want to use.
    ga('require', 'eventTracker', {
      attributePrefix: 'data-'
    });
    ga('require', 'outboundLinkTracker');
    ga('require', 'urlChangeTracker');
    // ...

    ga('send', 'pageview');
    </script>
    <script async src="https://www.google-analytics.com/analytics.js"></script>
    <script async src="autotrack.js"></script>


    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 

  </head>

  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li class="active"><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>
    </nav>


    <div id="front_page">

      <div class="col-md-10 col-md-offset-1">

        <div class="page-header">
          <h1>Relate</h1>
          <h4>Software to estimate genome-wide genealogies for thousands of samples</h4>
          <p><b>Relate</b> estimates genome-wide genealogies in the form of trees that adapt to changes in local ancestry caused by recombination. The method is scalable to thousands of samples. The method is described in the following paper. Please cite this paper if you use our software in your study.</p>   
          <p>
          <b>Citation:</b> Leo Speidel, Marie Forest, Sinan Shi, Simon Myers. A method for estimating genome-wide genealogies for thousands of samples. <a href="https://www.biorxiv.org/content/10.1101/550558v1">bioRxiv:550558</a>, 2019.</b>
          </p>          
          <p>
          <b>Contact:</b> leo.speidel [at] outlook [dot] com <b>Website:</b> <a href = "https://leospeidel.wordpress.com">https://leospeidel.wordpress.com</a>
          </p>
        </div> 

      </div>



      <div class="container-fluid">
        <div class="col-md-10 col-md-offset-1">

          <!-- 1st row -->
          <div class="row">

            <div class="col-md-8">

              <div class="panel panel-info">
                <div class="panel-body">  
                  <img src="figures/treeplot.jpg" >
                </div>
              </div>

            </div>

            <div class="col-md-4">

              <div class="row">
                <h4>Download </h4>

                <p><b>Relate</b> is available for academic use. To see rules for non-academic use, please read the <a href="download/20180411_15670_RELATE_Academic_Usage_Licence.pdf" download>LICENCE</a> file, which is included with each software download.</p>
               
                <h5>Pre-compiled binaries (last updated: 25/02/2019)</h5> 
                <form role="form" data-toggle="validator" action="download/relate_v1_linux.tar">
                  <div class="form-group">
                    <div class="checkbox">
                      <label>
                        <input type="checkbox" id="terms" oninvalid="this.setCustomValidity('Required')" onchange="this.setCustomValidity('')" required>
                        I agree with the <a href="download/20180411_15670_RELATE_Academic_Usage_Licence.pdf">terms and conditions</a>
                      </label>
                      <div class="help-block with-errors"></div>
                    </div>
                  </div>
                  <div class="form-group">
                    <button type="submit" 
                            class="btn btn-link" 
                            formaction="download/relate_v1.0.7_x86_64_dynamic.tgz"   
                            data-on="click"
                            data-event-category="Relate download (x86_64, dynamic)"
                            data-event-action="click">Linux (x86_64, dynamic) - v1.0.7</button><br/>
                    <button type="submit" 
                            class="btn btn-link" 
                            formaction="download/relate_v1.0.7_x86_64_static.tgz"
                            data-on="click"
                            data-event-category="Relate download (x86_64, static)"
                            data-event-action="click">Linux (x86_64, static) - v1.0.7</button><br/>
                    <button type="submit" 
                            class="btn btn-link" 
                            formaction="download/relate_v1.0.7_MacOSX.tgz"
                            data-on="click"
                            data-event-category="Relate download (Mac_OSX)"
                            data-event-action="click">Mac OSX - v1.0.7</button>                   
                  </div>
                </form>

                <p>In the downloaded directory, we have included a toy data set. You can try out <b>Relate</b> using this toy data set by following the instructions on our <a href="getting_started.html">getting started</a> page.</p>

                <p>If you have any problems getting the program to work on your machine or would like to request an executable for a platform not shown here, please send a message to leo.speidel [at] outlook [dot] com. </p>

                <p>We document changes to previous versions in a <a href="#change-log">change-log</a>.</p>
              </div>

            </div>

          </div>
          <!-- 2nd row -->
          <div class="row">
            <h3>Add-on modules</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="input_data.html">Preparing input files</a></h4></div>
                <div class="panel-body">  
                  The input file format of <b>Relate</b> is the haps/sample file format used as an output format of ShapeIt. We provide code to convert files from the hap/legend/sample and vcf file formats. <b>Relate</b> also comes with tools for determining the ancestral allele using an ancestral genome and tools for filtering SNPs using a genomic mask.    
                </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Extract">Extracting subtrees and trees of interest</a></h4></div>
                <div class="panel-body">  
                  We provide code for converting individual trees of interest into .newick format. This file format is convenient, for instance, when visualising a tree. We also provide code for extracting genealogies corresponding to a subpopulation.
                </div>
              </div>
            </div>

          </div>



          <!-- 4th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#TreeView">Plot marginal trees</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/tree.jpg">
                  <b>Relate</b> can be used to plot a marginal tree of interest. This tree corresponds to the LCT region, where a mutation at SNP rs4988235 is believed to be responsible for Lactose tolerance in Europeans (here GBR). We can see that the derived allele at this SNP has spread rapidly in GBR, which is indicative of strong positive selection. </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#CoalescenceRate">Estimating population sizes and separation histories</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/population_size.jpg">
                  <b>Relate</b> can be used to estimate population sizes and separation histories between populations. The figure shows the separation history between FIN and GBR in the 1000GP data set. The inset shows the matrix of coalescent rates between pairs of haplotypes 9,000 years ago. Rows and columns are sorted by population labels of haplotypes, as indicated by the colour on the left of the matrix. </div>
              </div>
            </div>

          </div>

          <!-- 5th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#MutationRate">Estimating mutation rates</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/mutation_rate.jpg">
                  <b>Relate</b> can be used to estimate mutation rates through time. The figure shows TCC to TTC mutation rates for all 26 populations in the 1000GP data set. Trends shared between mutation categories are eliminated by dividing by the overall average mutation rate. For each population, the mutation rate is normalized such that its mean over time equals one. Consistent with previous estimates, we see an increase in the mutation rate of TCC to TTC mutations about 10,000 to 30,000 years ago in Europeans and Southern Asians.</div>
              </div>
            </div>


            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Selection">Detecting positive selection</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/selection.jpg">
                  <b>Relate</b> can be used to detect evidence for positive selection. We calculate a p-value for selection evidence that quantifies how quickly a mutation has spread in the population. The figure shows a manhattan plot for GBR which indicates clear peaks in the LCT and MHC regions, which are known targets of positive selection.</div>
              </div>
            </div>



          </div>


        </div>
      </div>


     <div class="container-fluid">

   
        <!-- 1st row -->
        <div class="row">
          <div class="col-md-10 col-md-offset-1">

            <h3>Change-log</h3>
            <hr/>

            <a class="anchor" name="change-log"></a> 

            <div class="panel-group" id="accordion">

             <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse9">
                      v1.0.8
                    </a>
                  </h4>
                </div>
                <div id="collapse9" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 15th March 2019<br/>     
                    <ul>
                      <li>Now requiring different input and output names for DetectSelection.sh when it overwrites files otherwise.</li>
                      <li>DetectSelection.sh now works also if additional columns are not appended to the .mut file using ./RelateFileFormats --mode GenerateSNPAnnotations. However, code is more efficient if these columns are appended.</li>
                      <li>Switched to using R package cowplot in TreeView.sh.</li>
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse8">
                      v1.0.7
                    </a>
                  </h4>
                </div>
                <div id="collapse8" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 25th Feburary 2019<br/>     
                    <ul>
                      <li>Fixed bug in the painting which affects tree topologies sometimes, but has only a small effect (not visible in distance measures when comparing to truth in simulations).</li> 
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse7">
                      v1.0.6
                    </a>
                  </h4>
                </div>
                <div id="collapse7" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 6th Feburary 2019<br/>     
                    <ul>
                      <li>I added a function for converting anc/mut output files to tree sequence file format (tskit). Currenty, some information is lost by this conversion.</li> 
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse6">
                      v1.0.5
                    </a>
                  </h4>
                </div>
                <div id="collapse6" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 18th October 2018<br/>     
                    <ul>
                      <li>I had introduced a bug in update v1.0.4 in the function RelateExtract --mode TreeAtSNPAsNewick which has been fixed.</li> 
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse5">
                      v1.0.4
                    </a>
                  </h4>
                </div>
                <div id="collapse5" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 13th October 2018<br/>     
                    <ul>
                      <li>Substituted zcat with gunzip -c to fix a bug in Macs.</li>
                      <li>Added requirement of R version >= 3.3.1 for TreeView because of a known bug with grid.draw() in older versions.</li>
                      <li>Fixed bug in RelateExtract --mode TreeAtSNPAsNewick, which outputted the first tree when snp_of_interest was not a SNP in the data set. Changed option snp_of_interest to pos_of_interest and this function now prints the tree at the position of interest. In addition, output filename was not using the option -o, which has been corrected.</li> 
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse4">
                      v1.0.3
                    </a>
                  </h4>
                </div>
                <div id="collapse4" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 30th August 2018<br/>     
                    <ul>
                      <li>Implemented pipeline for calculating p-values for selection evidence. Updated corresponing entry in the documentation.</li>
                      <li>Bug fix in RelateSelection --mode Frequency: Previous version had a bug whenever two internal nodes had same age.</li>
                      <li>Fixed function RelateFileFormats --mode ConvertFromVcf which previously parsed the vcf incorrectly in some cases.</li>
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse3">
                      v1.0.2
                    </a>
                  </h4>
                </div>
                <div id="collapse3" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 16th July 2018<br/>     
                    <ul>
                      <li>Implemented parallelization of module EstimatePopulationSize.</li>
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse2">
                      v1.0.1
                    </a>
                  </h4>
                </div>
                <div id="collapse2" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 30th June 2018<br/>
                    <ul>
                      <li>Bug fix in parsing function of haps/sample.</li>
                    </ul>
                  </div>
                </div>
              </div>

              <div class="panel panel-default">
                <div class="panel-heading">
                  <h4 class="panel-title">
                    <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion" href="#collapse1">
                      v1.0.0
                    </a>
                  </h4>
                </div>
                <div id="collapse1" class="panel-collapse collapse">
                  <div class="panel-body">
                    Date: 4th June 2018
                    <ul>
                      <li> Initial release.</li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>

          </div>
        </div>
      </div>

    </div>

  </body>
</html>        

