<!DOCTYPE html>
<html lang="en">

  <head>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-115753061-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'UA-115753061-1');
    </script>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 

  </head>

  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li class="active"><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>
    </nav>


    <div id="front_page">

      <div class="col-md-10 col-md-offset-1">

        <div class="page-header">
          <h1>Relate</h1>
          <h4>Software to estimate genome-wide genealogies with recombination for thousands of samples</h4>
        </div> 

      </div>



      <div class="container-fluid">
        <div class="col-md-10 col-md-offset-1">

          <!-- 1st row -->
          <div class="row">

            <div class="col-md-8">

              <div class="panel panel-info">
                <div class="panel-body">  
                  <img src="figures/treeplot.jpg" >
                </div>
              </div>

            </div>

            <div class="col-md-4">
              <div class="row">
                <h4>Updates: </h4>
                <ul>
                  <li>...</li>

                  <li>...</li>
                </ul>
              </div>

              <div class="row">
                <h4>Download </h4>

              </div>

            </div>

          </div>

          <div class="row">
            <div class="col-md-12">
              <p><b>Relate</b> estimates genome-wide genealogies in form of trees that adapt to changes in the local genealogy caused by recombination. It is scalable to thousands of samples. The method is described in the following paper. Please cite this paper if you use our software in your study.</p>   
              <p>
              <b>Citation:</b> Leo Speidel, Simon Myers. A method for estimating genome-wide genealogies for thousands of samples. 2018.</b>
              </p>          
              <p>
              <b>Contact:</b> leo.speidel [at] outlook [dot] com
              </p>
            </div>          
          </div>

          <!-- 2nd row -->
          <div class="row">
            <h3>Add-on modules</h3>
            <hr/>
          </div>

          <!-- 3rd row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="input_data.html">Preparing input files</a></h4></div>
                <div class="panel-body">  
                  The input file format of <b>Relate</b> is the haps/sample file format used as an output format of ShapeIt. We provide code to convert files from the hap/legend/sample and vcf file formats. <b>Relate</b> also comes with a tool that determines the ancestral allele using an ancestral genome and a tool that filters SNPs using a genomic mask.    
                </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Extract">Extracting subtrees and trees of interest</a></h4></div>
                <div class="panel-body">  
                  We provide code for converting individual trees of interest into .newick format. This file format is convenient, for instance, when visualising a tree. We also provide code for extracting genealogies corresponding to a subpopulation.
                </div>
              </div>
            </div>

          </div>



          <!-- 4th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#TreeView">Plot tree of interest</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/tree.jpg">
                  <b>Relate</b> can be used to plot a marginal tree of interest. </div>
              </div>
            </div>

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#CoalescenceRate">Estimating population sizes and separation histories</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/population_size.jpg">
                  <b>Relate</b> can be used to estimate population sizes and separation histories between populations. The figure shows the separation history between FIN and GBR in the 1000GP data set. The inset shows the matrix of coalescent rates between pairs of haplotypes 9,000 years ago. Rows and columns are sorted by population labels of haplotypes, as indicated by the color on the left of the matrix. </div>
              </div>
            </div>

          </div>

          <!-- 5th row -->
          <div class="row">

            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#MutationRate">Estimating mutation rates</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/mutation_rate.jpg">
                  <b>Relate</b> can be used to estimate mutation rates through time. The figure shows TCC to TTC mutation rates for all 26 populations in the 1000GP data set. Trends shared between mutation categories are eliminated by dividing by the overall average mutation rate. For each population, the mutation rate is normalized such that its mean over time equals one. Consistent with previous estimates, we see an increase in the mutation rate of TCC to TTC mutations about 10,000 to 30,000 years ago in Europeans and Southern Asians.</div>
              </div>
            </div>


            <div class="col-md-6">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4><a href="modules.html#Selection">Detecting positive selection</a></h4></div>
                <div class="panel-body">  
                  <img src="figures/selection.jpg">
                  <b>Relate</b> can be used to detect evidence for positive selection. We calculate a p-value for selection evidence that quantifies how quickly a mutation has spread in the population. The figure shows a manhattan plot for GBR which indicates clear peaks in the LCT and MHC regions which are known targets of positive selection.</div>
              </div>
            </div>



          </div>


        </div>
      </div>

  </body>
</html>        

