<!DOCTYPE html>
<html lang="en">


  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 
  </head>



  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li class="active"><a href="preparing_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>

    </nav>


    <!-- CONTENT -->


    <div class="wrapper">

      <!-- Sidebar Holder -->
      <nav id="sidebar">

        <div class="sidebar-header">
          <h3>Overview</h3>
        </div>

        <ul class="list-unstyled components">
          <li>
            <a href="#FileFormat">Input file formats</a>
          </li>
          <li>
            <a href="#ConvertToFileFormat">Convert to haps/sample file format</a>
          </li> 
          <li>
            <a href="#Prepare">Preparing input files</a>
          </li> 
        </ul>

      </nav>

      <!-- Page Content Holder -->
      <div id="content">

        <div class="container-fluid">

          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">


              <a class="anchor" name="FileFormat"></a>
              <h3>Input data file format</h3>

              <p>The input file format for <b>Relate</b> is the haps/sample file format, which is used as an output file format by SHAPEIT2 (see <a href="http://www.shapeit.fr/pages/m02_formats/hapssample.html">here</a>).</p>

              <div class="table-responsive">
                <table class="table">
                  <tbody>
                    <tr>
                      <a class="anchor" name="haps"></a> 
                      <td>.haps</td>
                      <td>The first five columns are:
                        <ol>
                          <li>Chromosome number [integer]</li>
                          <li>SNP ID [string]</li>
                          <li>SNP position [integer]</li>
                          <li>Ancestral allele [char]</li>
                          <li>Alternative allele [char]</li>
                        </ol>
                        Then, the successive column pairs correspond to the haplotypes carried by an individual at that SNP.<br/>
                        <b>Please make sure that the ancestral allele is represented by 0 and the alternative allele is represented by 1.</b>
                        <pre><code>7 SNP1 123 A G 0 0 1 0 0 0 1 1<br/>7 SNP2 456 T C 0 1 1 0 0 1 0 1<br/>7 SNP3 789 A T 0 1 1 0 1 1 1 1</code></pre>
                        This file is space delimited.
                      </td>
                    </tr>
                    <tr>
                      <a class="anchor" name="sample"></a> 
                      <td>.sample</td>
                      <td>The three columns are:
                        <ol>
                          <li>First individual ID</li>
                          <li>Second individual ID</li>
                          <li>Missing data proportion</li>
                        </ol>
                        <pre><code>ID_1 ID_2 missing<br/>0    0    0<br/>UNR1 UNR1 0<br/>UNR2 UNR2 0<br/>UNR3 UNR3 0<br/>UNR4 UNR4 0<br/></code></pre>
                        This file is space delimited. 
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>

              <p>Recombination rates are specified using a genomic map.</p>
              <div class="table-responsive">
                <table class="table">
                  <tbody>
                    <tr>
                      <a class="anchor" name="map"></a> 
                      <td>.map</td>
                      <td>This is the standard file format for genetic recombination maps. The three columns are:
                        <ol>
                          <li>Position (b) [integer]</li>
                          <li>Recombination rate (cM/Mb) [float]</li>
                          <li>Genetic position (cM) [float]</li>
                        </ol>
                        <pre><code>position COMBINED_rate(cM/Mb) Genetic_Map(cM)<br/>0        2.6858 0.4103<br/>2529     2.8222 0.4174<br/>2601     2.9813 0.4176</code></pre>
                        This file is space delimited.
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>

              <p>For some add-on modules of <b>Relate</b>, we also want to specify population labels. The .poplabels file format is identical to the sample file in the <a href="http://www.shapeit.fr/pages/m02_formats/haplegsample.html">hap/legend/sample file format</a>.</p>
              <div class="table-responsive">
                <table class="table">
                  <tbody>
                    <tr>
                      <a class="anchor" name="poplabels"></a> 
                      <td>.poplabels</td>
                      <td>This is the standard file format for genetic recombination maps. The three columns are:
                        <ol>
                          <li>Individual ID as specified in the .sample file [string]</li>
                          <li>Population label [string]</li>
                          <li>Group label [string]</li>
                          <li>Sex [integer]</li>
                        </ol>
                        <pre><code>sample population group sex<br/>UNR1 PJB SAS 1<br/>UNR2 JPT EAS 2<br/>UNR3 GBR EUR 2<br/>UNR4 YRI AFR 2</code></pre>
                        This file is space delimited.
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>




            </div>
          </div>

          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">

              <a class="anchor" name="ConvertToFileFormat"></a>
              <h3>Convert to haps/sample file format</h3>
              <hr/>

            </div>
          </div>



          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">

              <a class="anchor" name="ConvertFromHapLegendSample"></a> 
              <div class="panel panel-info">
                <div class="panel-heading"> <h4>Convert from hap/legend/sample</h4></div>
                <div class="panel-body">  
                  Converts from the hap/legend/sample file format (used for reference panels in e.g., IMPUTE2, see <a href="http://www.shapeit.fr/pages/m02_formats/haplegsample.html">here</a>) to the haps/sample file format. The code deletes non-biallelic SNPs.
                  <ul>
                    <li>Arguments --haps and --sample are used to specify output filenames.</li>
                    <li>Argument -i is used to specify the input filename (without file extension).</li>
                    <li>Use argument --chr to specify a chromosome index [int] (used as the first column in the haps file).</li>
                  </ul>

                  <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode ConvertFromHapLegendSample \
                 --haps example.haps \
                 --sample example.sample \
                 -i example \
                 --chr 1  </code></pre>
                  Output: example.haps, example.sample.

                </div>
              </div>
            </div>
          </div>

          <div class="row">

            <div class="col-md-10 col-md-offset-1">

              <a class="anchor" name="ConvertFromVcf"></a> 
              <div class="panel panel-info">
                <div class="panel-heading"> <h4>Convert from vcf</h4></div>
                <div class="panel-body">  

                  Converts from the vcf file format to the haps/sample file format.
                                   
                  <ul>
                    <li>Arguments --haps and --sample are used to specify output filenames.</li>
                    <li>Argument -i is used to specify the input filename (without file extension).</li>
                  </ul>

                  <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode ConvertFromVcf \
                 --haps example.haps \
                 --sample example.sample \
                 -i example </code></pre>
                  Output: example.haps, example.sample.

                </div>
              </div>
            </div>

          </div>
        </div>

        <!-- row -->
        <div class="row">

          <a class="anchor" name="RemoveNonBiallelicSNPs"></a> 
          <div class="col-md-10 col-md-offset-1">


            <a class="anchor" name="Prepare"></a>
            <h3>Preparing input files</h3>
            <hr/>
            <p>The following code can be used to prepare the input data. We recommend to apply all steps (if applicable) in the listed order. 
            </p>

            <div class="panel panel-info">
              <div class="panel-heading"> <h4>Remove non-biallelic SNPs</h4></div>
              <div class="panel-body">  

                Removes SNPs that are not biallelic in the data set.
                <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode RemoveNonBiallelicSNPs \
                 --haps example.haps \
                 --sample example.sample \
                 -o example_biallelic </code></pre>
                Output: example_biallelic.haps.

              </div>
            </div>
          </div>

        </div>


        <!-- row -->
        <div class="row">

          <a class="anchor" name="RemoveSamples"></a> 
          <div class="col-md-10 col-md-offset-1">
            <div class="panel panel-info">
              <div class="panel-heading"> <h4>Remove samples from data set</h4></div>
              <div class="panel-body">  

                Write ids of individuals as specified in the .sample file into remove_ids.txt (one id per line). These indivduals will be removed from the data set.
                <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode RemoveSamples \
                 --haps example.haps \
                 --sample example.sample \
                 -i remove_ids.txt \
                 -o example_subset </code></pre>
                Output: example_subset.haps, example_subset.sample.

              </div>
            </div>
          </div>

        </div>


        <!-- row -->
        <div class="row">

          <a class="anchor" name="FlipHapsUsingAncestor"></a> 
          <div class="col-md-10 col-md-offset-1">
            <div class="panel panel-info">
              <div class="panel-heading"> <h4>Determine ancestral allele and flip if necessary</h4></div>
              <div class="panel-body">  

                The ancestral genome should be saved as a fasta file and aligned to the data set. The code is insensitive to case in the fasta file and deletes SNPs at which neither allele matches the ancestral allele.
                <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode FlipHapsUsingAncestor \
                 --haps example.haps \
                 --sample example.sample \
                 --ancestor ancestor.fa \
                 -o example_ancestral </code></pre>
                Output: example_ancestral.haps

              </div>
            </div>
          </div>

        </div>


        <!-- row -->
        <div class="row">

          <a class="anchor" name="FilterHapsUsingMask"></a> 
          <div class="col-md-10 col-md-offset-1">
            <div class="panel panel-info">
              <div class="panel-heading"> <h4>Filter SNPs using a genomic mask</h4></div>
              <div class="panel-body">  

                The genomic mask should be saved as a fasta file and aligned to the data set. Denote the passing state by P. The code deletes all SNPs that are not passing. It also outputs a .dist file that contains the distances between SNPs, adjusted for regions that are not passing in the mask.
                <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode FilterHapsUsingMask\
                 --haps example.haps \
                 --sample example.sample \
                 --mask genomic_mask.fa \
                 -o example_mask </code></pre>
                Output: example_mask.haps, example_mask.dist.

              </div>
            </div>
          </div>

        </div>



        <!-- row -->
        <div class="row">

          <a class="anchor" name="GenerateSNPAnnotations"></a> 
          <div class="col-md-10 col-md-offset-1">
            <div class="panel panel-info">
              <div class="panel-heading"> <h4>Generate SNP annotations</h4></div>
              <div class="panel-body">  

                This code uses the ancestral genome and population labels of samples to append additional information to the output of <b>Relate</b>. Fileformat for population labels is described <a href="preparing_data.html#poplabels">here</a>. The ancestral genome should be saved in fasta file format and aligned to the data set. Argument --ancestor is optional. If argument --mut is specified, output is directly appended to the .mut file.</br>
                        <pre><code> PATH_TO_RELATE/bin/RelateFileFormats \
                 --mode GenerateSNPAnnotations \
                 --haps example.haps \
                 --sample example.sample \
                 --poplabels example.poplabels \
                 --ancestor ancestor.fa \
                 -o example </code></pre>
                        Output: example.annot. If --mut is specified, columns are appended to this file.

              </div>
            </div>
          </div>

        </div>

      </div>

    </div>

    </div>


  </body>
</html>     
